chaincode/marbles_transfer/marbles_transfer_queries.go

```diff
@@ -106,16 +114,78 @@ func queryMarblePrice(ctx contractapi.TransactionContextInterface, marbleID stri
+// QueryAssetSaleAgreements returns all of an organization's proposed sales
+func (s *SmartContract) QueryAssetSaleAgreements(ctx contractapi.TransactionContextInterface) ([]Agreement, error) {
+	return queryAgreementsByType(ctx, typeMarbleForSale)
+}
+
+// QueryAssetBuyAgreements returns all of an organization's proposed buys
+func (s *SmartContract) QueryAssetBuyAgreements(ctx contractapi.TransactionContextInterface) ([]Agreement, error) {
+	return queryAgreementsByType(ctx, typeMarbleBid)
+}
+
+func queryAgreementsByType(ctx contractapi.TransactionContextInterface, agreeType string) ([]Agreement, error) {
+	collection, err := getClientImplicitCollectionName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query for any object type starting with `agreeType`
+	agreementsIterator, err := ctx.GetStub().GetPrivateDataByPartialCompositeKey(collection, agreeType, []string{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from private data collection: %s", err.Error())
+	}
+	defer agreementsIterator.Close()
+
+	agreements := []Agreement{}
+
+	for agreementsIterator.HasNext() {
+		resp, err := agreementsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		newAgree := new(Agreement)
+		err = json.Unmarshal(resp.Value, newAgree)
+		if err != nil {
+			return nil, err
+		}
+
+		agreements = append(agreements, *newAgree)
+	}
 
-// TODO add a query to get all of an organization's proposed buys
-// Use GetPrivateDataByPartialCompositeKey to find all keys starting with typeMarbleBid
+	return agreements, nil
+}
 
 // TODO add a JSON index and query to return all of an organization's marbles larger than a certain size (only works when using CouchDB state database)
 // hint: see sample index at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/META-INF/statedb/couchdb/indexes/indexOwner.json
 // hint: see sample query at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L515
 
-// TODO add a history query so that users can see the chain of custody for a marble since issuance
-// hint: see sample at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L692
+// QueryAssetHistory returns the chain of custody for a marble since issuance
+func (s *SmartContract) QueryAssetHistory(ctx contractapi.TransactionContextInterface, marbleID string) ([]QueryResult, error) {
+	resultsIterator, err := ctx.GetStub().GetHistoryForKey(marbleID)
+	if err != nil {
+		return nil, err
+	}
+	defer resultsIterator.Close()
+
+	records := []QueryResult{}
+
+	for resultsIterator.HasNext() {
+		response, err := resultsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		marble := new(Marble)
+		err = json.Unmarshal(response.Value, marble)
+		if err != nil {
+			return nil, err
+		}
+
+		record := QueryResult{Record: marble}
```

Comment:
We'll want the history results to look similar to the JSON built up at:
https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L714-L748

That is, the user that is querying for the history of the asset changes will likely want to see the txid and timestamp of the changes (IsDelete is not relevant in this chaincode).

(but use Unmarshal() instead of writing the JSON manually... that needs to be changed in marbles02)